packages/extension/src/background/transaction/actions.ts

```diff
@@ -20,95 +20,146 @@ import { KeyregTransaction } from "./keyregTransaction";
+          validatedTxnWrap = new AssetCreateTransaction(txn as IAssetCreateTx);
+        } catch (e) {
+          e.message = [error.message, e.message].join(' ');
+          error = e;
+        }
+      }
+      if (!validatedTxnWrap) {
+        try {
+          validatedTxnWrap = new AssetDestroyTransaction(
+            txn as IAssetDestroyTx
+          );
+        } catch (e) {
+          e.message = [error.message, e.message].join(' ');
+          error = e;
+        }
+      }
+      if (!validatedTxnWrap && error) {
+        throw error;
+      }
+      break;
+    case TransactionType.Afrz:
+      validatedTxnWrap = new AssetFreezeTransaction(txn as IAssetFreezeTx);
+      break;
+    case TransactionType.Axfer:
+      // Validate any of the 3 types of transactions that can occur with axfer
+      // Use the first error as the passback error.
+      try {
+        validatedTxnWrap = new AssetAcceptTransaction(txn as IAssetAcceptTx);
+      } catch (e) {
+        error = e;
+      }
+      if (!validatedTxnWrap) {
+        try {
+          validatedTxnWrap = new AssetTransferTransaction(
+            txn as IAssetTransferTx
+          );
+        } catch (e) {
+          e.message = [error.message, e.message].join(' ');
+          error = e;
+        }
+      }
+      if (!validatedTxnWrap) {
+        try {
+          validatedTxnWrap = new AssetClawbackTransaction(
+            txn as IAssetClawbackTx
+          );
+        } catch (e) {
+          e.message = [error.message, e.message].join(' ');
+          error = e;
+        }
+      }
+      if (!validatedTxnWrap && error) {
+        throw error;
+      }
+      break;
+    case TransactionType.Keyreg:
+      validatedTxnWrap = new KeyregTransaction(txn as IKeyRegistrationTx);
+      break;
+    case TransactionType.Appl:
+      validatedTxnWrap = new ApplTransaction(txn as IApplTx);
+      break;
+    default:
+      throw new Error('Type of transaction not specified or known.');
+  }
+
+  return validatedTxnWrap;
+}
 
-    return validatedTxnWrap;
+export function getLedgerFromGenesisID(genesisID: string):Ledger {
+    let ledger;
+    if (genesisID === 'mainnet-v1.0') ledger = Ledger.MainNet;
+    else if (genesisID === 'testnet-v1.0') ledger = Ledger.TestNet;
+    return ledger;
 }
 
+export function calculateEstimatedFee(
+  transactionWrap: BaseValidatedTxnWrap,
+  params: any
+): void {
+  const transaction = transactionWrap.transaction;
+  const minFee = +params['min-fee'];
+  let estimatedFee = +transaction['fee'];
+  if (transaction['flatFee']) {
+    // If flatFee is enabled, we compare the presented fee by the dApp against the
+    // Ledger suggested min-fee and use the min-fee if higher than the presented fee
+    if (estimatedFee < minFee) {
+      estimatedFee = minFee;
+    }
+  } else {
+    let dappFee = estimatedFee;
+    // If the dApp doesn't suggest a per-byte fee, we use the Ledger suggested min-fee
+    if (dappFee === 0) {
+      dappFee = minFee;
```

Comment:
I might be wrong, but shouldn't this be the final fee when the non flat fee is 0? I mean, just use the minFee and don't multiply it by the byte size?

Reply:
I just tested it and you appear to be right